save settings in json instead of xml

akashters opened the pull request:
- use nlohmann_json library to save settings in json format instead of xml format
- remove boost dependency since it's only being used for xml parsing and no longer needed

akashters committed:
use nlohmann_json and save settings in json instead of xml

settings are currently being saved in xml format using boost
property_tree library, this change uses nlohmann_json library to save
the settings in json format instead.

akashters committed:
remove boost dependency

boost is no longer needed since property tree dependency is replaced
by nlohmann_json library

shriramters commented on src/main_screen.cpp:
```diff
@@ -1,26 +1,27 @@
 #include <string>
 #include <vector>
 
+#define SETTINGS_FILE_NAME "settings.json"
```
Use typed constexpr

```c++
constexpr std::string_view settings_file_name = "settings.json";
```

akashdoppalapudi replied:
Thanks. I replaced #define with a constexpr

shriramters commented on src/preferences.hpp:
```diff
@@ -38,7 +37,16 @@ struct Preferences {
     void save(const std::string& filename);
 };
+
+class SettingsFileDoesntExistException : public std::exception {
+  private:
+    std::string message;
+
+  public:
+    SettingsFileDoesntExistException() : message("Settings file doesn't exist at the given path") {}
+
+    const char* what() const throw() { return message.c_str(); }
+};
```
Personally I would keep it simple and throw `std::runtime_error("file doesn't exist")`

akashdoppalapudi replied:
I used a runtime_error and removed the custom exception, also moved the preferences related code to preferences.cpp thus removing clutter from the main_screen.cpp

shriramters commented on src/utils.hpp:
```diff
@@ -4,6 +4,7 @@
 #include "chord_db.hpp"
 #include <algorithm>
 #include <iostream>
+#include <optional>
```
good catch

shriramters commented on CMakeLists.txt:
Is it necessary to format CMakeLists like this? is this being done by clang format?

My flatpak manifest applies a patch on this which will have to be completely rewritten now if this is formatted

akashdoppalapudi replied:
I guess it's okay to keep it this way

akashters committed:
move preferences related code entirely to preferences.cpp

- move preferences related code to preferences.cpp and clear up some
clutter in main_screen.cpp
- make methods inside utils.hpp inline so that they can be used in
multiple files

akashdoppalapudi commented:
I also think it's better to separate declarations and implementations in utils.hpp with a utils.cpp file.
We should consider it.

shriramters reviewed (approved):
LGTM 
Great work
and also, thanks
